swarm/src/behaviour.rs

```diff
@@ -721,3 +838,214 @@ impl Default for CloseConnection {
         CloseConnection::All
     }
 }
+
+pub enum InEvent<'a, Handler: IntoConnectionHandler> {
+    /// Informs the behaviour about a newly established connection to a peer.
+    ConnectionEstablished {
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        endpoint: &'a ConnectedPoint,
+        // TODO: Would a slice not be better?
+        failed_addresses: Option<&'a Vec<Multiaddr>>,
+        other_established: usize,
+    },
+    /// Informs the behaviour about a closed connection to a peer.
+    ///
+    /// This event is always paired with an earlier
+    /// [`InEvent::ConnectionEstablished`] with the same peer ID, connection ID
+    /// and endpoint.
+    ConnectionClosed {
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        endpoint: &'a ConnectedPoint,
+        handler: <Handler as IntoConnectionHandler>::Handler,
+        remaining_established: usize,
+    },
+    /// Informs the behaviour that the [`ConnectedPoint`] of an existing
+    /// connection has changed.
+    AddressChange {
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        old: &'a ConnectedPoint,
+        new: &'a ConnectedPoint,
+    },
+    /// Informs the behaviour about an event generated by the handler dedicated to the peer
+    /// identified by `peer_id`. for the behaviour.
+    ///
+    /// The `peer_id` is guaranteed to be in a connected state. In other words,
+    /// [`InEvent::ConnectionEstablished`] has previously been received with this `PeerId`.
+    ConnectionHandler {
+        peer_id: PeerId,
+        connection: ConnectionId,
+        event: <<Handler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent,
+    },
```

Comment:
Referencing a suggestion from @thomaseizinger  in #2854: what do you think about splitting this into two methods `on_swarm_event` and `on_handler_event`? 
Imo this would make it easier to understand the kind of event, its source, and how to handle it. 
One major difference between the two types of events are for example that events from the handler are something that the behaviour should definitely handle and that differ depending on the type of behaviour, whereas swarm events are more "informative" in the sense that the behaviour may handle it, but doesn't have to if they are not relevant.

Reply:
I am also in favor of my suggestion :relaxed:

Reply:
Done via https://github.com/libp2p/rust-libp2p/pull/2867/commits/07a4f201b78978cb56a6f24fd1426de5664447cc.